### PR TITLE
Modified mask weights functionality

### PR DESCRIPTION
**Pull request by MattsonCam**
Added the ability to compute the exponential moving average of the mask weights in the BCE loss function. Also removed the ability to specify custom mask weights.

### CODE REVIEW

File: metrics/BCE.py [hunk 1]
```diff
@@ -90,10 +88,24 @@ def forward(
             generated_predictions = generated_predictions.unsqueeze(0)
             targets = targets.unsqueeze(0)
 
+        # Exponential Moving Average of semantic mask pixel frequencies
+        if self.mask_weights_alpha is not None:
+            next_mask_weights = 1.0 / (targets.sum(dim=(0, 2, 3, 4)) + 1e-9)
```
Comment (wli51) [inline metrics/BCE.py]: i imagine the auto rolling-weighting of semantic classes by their inverse mean pixel intensity is meant to prevent the model from neglecting semantic classes with less/smaller objects? Should we expect the targets to always be binary and `targets.sum(...) > 1` and never 0? If not, then having a 1.0/[small eps] could result in an exploded weight. I think for your purpose doing `1.0 / (targets.sum(dim=(0, 2, 3, 4)) + 1)` would be a safer and smoother way of getting the inverse weights based on pixel and you will also be able to compute weights from normalized stains (which might not be useful though).   

In addition to that, is there any chance in semantic segmentation and patched datasets to have target patches where only one semantic class had masks and the other ones are just empty? If so, you might want to skip completely empty channels from the rolling mean weight computation?

File: metrics/BCE.py [hunk 2]
```diff
@@ -38,15 +36,15 @@ def __init__(
...
             )
 
-        self.mask_weights_sum = self.mask_weights.sum()
-
         self.is_loss = is_loss
         self.use_logits = use_logits
         self.device = (
             device if isinstance(device, torch.device) else torch.device(device)
         )
+        self.mask_weights_alpha = mask_weights_alpha
```
Comment (wli51) [inline metrics/BCE.py:53]: Since you are not doing any checks here and i see from your optuna trails that this gets specified by the sampler, it is the safest to cast all weights explicitly to the fp32 dtype (or dtype of your model weights). I think torch automatically casts numpy values to the equivalent tensor precision and in multiplication the lower precision tenor to match the higher precision tensor. 

In the event that the `mask_weights_alpha` happened to be `np.float64`, multiplying them with your `torch.ones(3, dtype=torch.float32)` weights would make the weights `torch.float64` and can cause downstream errors if it hits any layer/metric/loss that does not like different precisions for input/target/nn module weights. Since you are weighting after the the BCE loss you shouldn't get any precision errors, though if you `loss.backward()` from a float64 loss I think torch by default computes the gradients in 64bit precision too, which is expensive (twice as slow on data center grade gpus like A100 and H100s and many times more slower on consumer grade cards we have) and not useful if it gets applied back to float32 models.

**Review by wli51 — APPROVED** [review]
LGTM! Please see my comments for minor things.

File: metrics/BCE.py (hunk 1 above)
Comment (MattsonCam) [reply metrics/BCE.py]: The BCE loss should be zero for that semantic mask if it contains no positive pixels. However, the weight will still matter, since the weights are normalized. If one semantic mask's weight is very large, then the other masks normalized weights could be very small, so the total weighted normalized loss would be small.

It might be worth including some sort of threshold in case only a few pixels of a semantic mask are present. That way I could make its weight smaller compared to the other weights, but this is what the inverse relationship is supposed to solve (except maybe for the case when there are no positive pixels in the semantic mask). I might also just make the weight very small if no positive pixels from a semantic mask present.

File: metrics/BCE.py (hunk 2 above)
Comment (MattsonCam) [reply metrics/BCE.py:53]: Hadn't thought about this, but it is good to know. Do you know how this behaves with torch AMP?

Comment (wli51) [reply metrics/BCE.py:53]: If torch auto-casts the weight alpha/weight to float32 or 16 thats probably fine. However, torch AMP is explicitly documented to not work with flaot64 [document page](https://docs.pytorch.org/docs/stable/amp.html#op-eligibility). My guess is that unless some casting is explicitly enforced the float64 intermediates will eventually hit some float32 or float16 module weights and result in an error.

Comment (MattsonCam) [reply metrics/BCE.py:53]: Thanks for bringing this to my attention. I think I addressed this

**Comment by MattsonCam** [comment]
Thanks for the review @wli51, merging now!
